include/psa/crypto.h

```diff
@@ -4297,63 +4184,218 @@ static psa_pake_operation_t psa_pake_operation_init(void);
+ *                              must have been set up by psa_pake_setup() and
+ *                              not yet in use.
  * \param password              Identifier of the key holding the password or a
  *                              value derived from the password (eg. by a
  *                              memory-hard function).  It must remain valid
  *                              until the operation terminates. It must be of
- *                              type PSA_KEY_TYPE_PASSWORD or
+ *                              type #PSA_KEY_TYPE_PASSWORD or
  *                              #PSA_KEY_TYPE_DERIVE. It has to allow the usage
  *                              #PSA_KEY_USAGE_DERIVE.
- * \param cipher_suite          The cipher suite to use. (A cipher suite fully
- *                              characterizes a PAKE algorithm and determines
- *                              the algorithm as well.)
- * \param side                  A value of type ::psa_pake_side_t signaling the
- *                              side of the algorithm that is being set up. For
- *                              more information see the documentation of \c
- *                              PSA_PAKE_SIDE_XXX constants.
- * \param[in] user_id           The user ID to authenticate with.
- * \param user_id_len           Size of the \p user_id buffer in bytes.
- * \param[in] peer_id           The peer's ID to authenticate.
- * \param peer_id_len           Size of the \p peer_id buffer in bytes.
- * \param[in] session_data      Additional session related data if it is allowed
- *                              or required by the algorithm. This must be empty
- *                              if additional session data is not used by the
- *                              algorithm.
- * \param session_data_len      Size of the \p session_data buffer in bytes.
  *
  * \retval #PSA_SUCCESS
  *         Success.
  * \retval #PSA_ERROR_BAD_STATE
- *         The operation state is not valid (it must be inactive).
+ *         The operation state is not valid (it must have been set up.)
+ * \retval #PSA_ERROR_CORRUPTION_DETECTED
  * \retval #PSA_ERROR_INVALID_HANDLE
+ * \retval #PSA_ERROR_COMMUNICATION_FAILURE
+ * \retval #PSA_ERROR_HARDWARE_FAILURE
+ * \retval #PSA_ERROR_STORAGE_FAILURE
  * \retval #PSA_ERROR_NOT_PERMITTED
  * \retval #PSA_ERROR_INVALID_ARGUMENT
  *         \p key is not compatible with the algorithm in \p cipher_suite.
+ * \retval #PSA_ERROR_BAD_STATE
+ *         The library has not been previously initialized by psa_crypto_init().
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_pake_set_password_key(psa_pake_operation_t *operation,
+                                       mbedtls_svc_key_id_t password);
+
+/** Set the password for a password-authenticated key exchange via a memory hard
+ * function.
```

Comment:
> We're using “key stretching” as the generic term for such functions.

Thanks, will do.

> I think a better name for this function would be `psa_key_derivation_output_for_pake`, with the key derivation operation parameter (the source) coming before the pake operation parameter (the destination).

Both of these are operation objects technically [in,out] parameters. `operation` takes an input from `key_derivation`, but  the state of `key_derivation` changes as well.

I think this function is better thought of as related to the PAKE operation, because it needs direct/low level access to the PAKE structure, whereas (in theory) it can use the public API of the key derivation operation without any need for low level access to the key derivation structure.

Reply:
After discussion during the Crypto call today, I am happy to go with Gilles suggestion to organise multi-operation APIs with input operations first, followed by output operations. This is considering the logical usage of the operation, not the `const`ness of the operation.

This ordering is systematically used for the non-operation parameters in all PSA Crypto functions, so it is consistent to continue that for APIs such as this.

I think this is a stronger argument for making this a `psa_key_derivation_xxx()` function than whether or not the implementation can use an existing published API to achieve the same effect.

`psa_key_derivation_output_for_pake()` is kind of clunky as a name, and isn't specific as to which part of the PAKE inputs are being provided, and PAKE takes a lot of them. Would `psa_key_derivation_output_pake_password()` be better?

Reply:
I don't like `psa_key_derivation_output_pake_password` because it makes it look like it's outputting the password. But the password is actually an input to the key derivation operation. What's being input into the PAKE is a value derived from the password.

Reply:
I finally took the time tot read the J-PAKE spec material, and the publically available THREAD material (which happens to be the mbedtls implementation in OpenThread :wink:).

The input to the J-PAKE algorithm is a [shared] low entropy secret _s_, which is described as being a password or hash of a password. J-PAKE does not require this to be a memory- or computationally- hard process - and that would potentially undermine THREAD's use of J-PAKE in IoT devices. A cryptographic hash of the password is suggested for J-PAKE in order to ensure that _s_ can be converted to an integer for use in the key establishment process without placing restrictions on the password itself to meet the J-PAKE requirements for _s_.

This leads to a couple of observations/questions:

1. The secret/password input to J-PAKE is unlikely to be a password-hash (PBKDF2-like), so this API is not that relevant for J-PAKE?
2. Should we use 'secret' in the interface names and description for this input to a PAKE, rather than 'password'? - we can describe the use of passwords, cryptographic hashes of passwords, or full password-hash in the full description, or in the individual algorithm descriptions?

Reply:
If this function is not useful for PAKE, and given that we aren't sure about its design, I'm for leaving it out.

> Should we use 'secret' in the interface names and description for this input to a PAKE, rather than 'password'?

We're already calling the output of a PAKE a “shared secret”. The password is a shared secret (for PAKE where the verifier knows a password-equivalent secret), but calling it “shared secret” or even just “secret” would be confusing.

Reply:
> If this function is not useful for PAKE, and given that we aren't sure about its design, I'm for leaving it out.

I was on the same opinion and it is already removed in the current version.

Reply:
(It was probably a typo, but just to clarify: it is useful for PAKE, just not for J-PAKE.)

Reply:
> We're already calling the output of a PAKE a “shared secret”. The password is a shared secret (for PAKE where the verifier knows a password-equivalent secret), but calling it “shared secret” or even just “secret” would be confusing.

Good point, I hadn't thought about that. Reverted the change.